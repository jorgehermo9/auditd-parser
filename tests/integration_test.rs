use std::{fs, path::PathBuf};

use auditd_parser::{AuditdRecord, ParserError};
use erased_serde::Serialize;
use rstest::rstest;

#[rstest]
fn test_log_data(#[files("tests/data/**/*.log")] log_file: PathBuf) {
    let log_file_path = log_file.to_string_lossy().into_owned();
    let log_file_content = fs::read_to_string(log_file).unwrap();
    let logs = log_file_content
        .lines()
        .enumerate()
        .map(|(index, line)| (index + 1, line))
        .filter(|(_line_number, line)| !line.is_empty() && !line.starts_with('#'))
        .collect::<Vec<_>>();

    for (line_number, log) in logs {
        let maybe_record = log.parse::<AuditdRecord>();
        let location = format!("{log_file_path}:{line_number}");
        let log_identifier = get_log_identifier(log, &location);

        // Unwrap the Ok variant but leave the Err variant as-is
        let result: Box<dyn Serialize> = match maybe_record {
            Ok(record) => Box::new(record),
            Err(error) => Box::new(Err::<AuditdRecord, ParserError>(error)),
        };

        insta::with_settings!(
            {
                info => &log,
                description => location,
                snapshot_suffix => log_identifier,
            },
            {
                insta::assert_json_snapshot!(result);
            }
        );
    }
}

fn get_log_identifier(log: &str, location: &str) -> String {
    let log_md5 = md5::compute(format!("{log}{location}"));
    format!("{log_md5:x}")
}
