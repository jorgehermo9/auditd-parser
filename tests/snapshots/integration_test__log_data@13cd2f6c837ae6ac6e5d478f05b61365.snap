---
source: tests/integration_test.rs
description: "/home/jorge/github/auditd-parser/tests/data/linux-audit-parser-rs/line-user-start.log:0"
expression: result
info: "type=USER_START msg=audit(1725004605.811:105663): pid=14548 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:session_open grantors=pam_keyinit,pam_keyinit,pam_limits,pam_systemd,pam_unix acct=\"root\" exe=\"/usr/bin/sudo\" hostname=? addr=? terminal=/dev/pts/0 res=success'"
---
{
  "record_type": "USER_START",
  "timestamp": 1725004605811,
  "id": 105663,
  "fields": {
    "auid": null,
    "msg": {
      "acct": "root",
      "addr": null,
      "exe": "/usr/bin/sudo",
<<<<<<< Updated upstream
      "grantors": "pam_keyinit,pam_keyinit,pam_limits,pam_systemd,pam_unix",
      "hostname": "?",
=======
      "grantors": [
        "pam_keyinit",
        "pam_keyinit",
        "pam_limits",
        "pam_systemd",
        "pam_unix"
      ],
      "hostname": null,
>>>>>>> Stashed changes
      "op": "PAM:session_open",
      "res": "success",
      "terminal": "/dev/pts/0"
    },
    "pid": "14548",
    "ses": "4294967295",
    "uid": "root"
  }
}
